pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java

```diff
@@ -99,11 +99,12 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
                 if (pulsarSourceConfig.getDeadLetterTopic() != null && !pulsarSourceConfig.getDeadLetterTopic().isEmpty()) {
                     deadLetterPolicyBuilder.deadLetterTopic(pulsarSourceConfig.getDeadLetterTopic());
                 }
-                cb.deadLetterPolicy(deadLetterPolicyBuilder.build());
+                cb = cb.deadLetterPolicy(deadLetterPolicyBuilder.build());
             }
 
-            return cb.subscribeAsync();
-        }).collect(Collectors.toList()).stream().map(CompletableFuture::join).collect(Collectors.toList());
+            Consumer<T> consumer = cb.subscribeAsync().join();
```

Comment:
This could be simplified into:
```suggestion
            Consumer<T> consumer = cb.subscribe();
```

In any case, I think we should be catching any exception thrown over the loop and close down all the consumers that were already created and added to the list.

Reply:
The current philosophy is that all cleaning up logic is inside the close which will get called as soon as there is an error. Shouldn't that suffice?

Reply:
The client close? What wouldn't work when running in the thread mode, since the `PulsarClient` is shared.